Fix archive pres copy creation (depends on #956, #954)

jmartin-sul opened the pull request:
ported and fixed the query, and added a test which would've demonstrated issue #899.

still need to:
- [x] finish porting `create_archive_preserved_copies!` to `PreservedCopy`
- [x] port/write tests for `PreservedCopy.create_archive_preserved_copies!`
- [x] write tests for `ArchiveEndpoint.which_have_archive_copy`
- [x] ~~remove `Endpoint.which_need_archive_copy` (will likely happen after rebasing on the below mentioned PRs anyway)~~ yup, 954 took care of it
- [x] rebase on master after #956 is merged
- [x] rebase on master after #954 is merged

i think the tentative plan is to get #956 merged, then #954, then this.

fixes #899

SaravShah commented:
cool `which_have` and `which_need` `archive_preserved_copies` verified working

jmartin-sul commented:
ready for review (@ndushay, @SaravShah, @tallenaz)

jmartin-sul commented on spec/models/preserved_copy_spec.rb:
```diff
@@ -309,4 +314,53 @@
     end
   end
+
+  describe '#create_archive_preserved_copies!' do
```
one thing i'd considered was writing tests for `#create_archive_preserved_copies!` that used two different druids, similarly to what i did for the `which_have_archive_copy` and `which_need_archive_copy` scopes on `ArchiveEndpoint`.  i'd still be open to do that, but felt like this was complete enough to be a PR as-is, especially since the scope that `#create_archive_preserved_copies!` relies on to figure out what to create is tested in that way.  figured it might be better to get this merged and ticket follow-on work.

tallenaz replied:
I agree it would be a good idea to have tests for `create_archive_preserved_copies!` using 2 different druids, and also agree that's fine for follow-on work. Otherwise LGTM.

jmartin-sul replied:
#986

jmartin-sul commented on app/models/archive_endpoint.rb:
```diff
@@ -16,6 +16,23 @@ class ArchiveEndpoint < ApplicationRecord
+      )
+  }
+
+  scope :which_need_archive_copy, lambda { |druid, version|
```
@SaravShah had suggested trying to work some bounds checking into the query, so that it doesn't return results for versions greater than the current version on the preserved object.  i'd like to try to do that, but it's not immediately clear to me what the query would be, so i have a slight preference for getting this merged and trying to do that as follow-on work.

if we did do that bounds checking in this query, it might or might not obviate the need for it in `PreservedCopy#create_archive_preserved_copies!`... i still sorta like the idea of failing noisily if an out of bounds version is given to that method, but i don't feel super-strongly about that.

SaravShah replied:
follow-on ticket sounds good to me.

jmartin-sul replied:
#983

ndushay commented on app/models/archive_endpoint.rb:
```diff
@@ -16,6 +16,23 @@ class ArchiveEndpoint < ApplicationRecord
+    joins(preservation_policies: [:preserved_objects]).where(preserved_objects: { druid: druid })
+  }
+
+  scope :which_have_archive_copy, lambda { |druid, version|
```
maybe add comment:
"for a druid, version, which zip_endpoints have  zipped_moab_versions  for the complete_moab?"

ndushay commented on app/models/archive_endpoint.rb:
```diff
@@ -16,6 +16,23 @@ class ArchiveEndpoint < ApplicationRecord
+      )
+  }
+
+  scope :which_need_archive_copy, lambda { |druid, version|
```
maybe add comment:

"for a druid, version, which zip_endpoints need a zipped_moab_version, per preservation_policy"

SaravShah commented on spec/models/archive_endpoint_spec.rb:
```diff
@@ -73,4 +73,82 @@
   end
+
+  describe '.archive_targets' do
+    let!(:alternate_pres_policy) do
```
do we want a factory for preservation policy?

jmartin-sul replied:
would be inclined to do that as another PR, but seems like a good idea

jmartin-sul replied:
#984

ndushay commented on spec/models/archive_endpoint_spec.rb:
```diff
@@ -73,4 +73,82 @@
+    end
+    let(:ma1_ep) { ArchiveEndpoint.find_by!(endpoint_name: 'mock_archive1') }
+
+    # TODO: try expecting to eq a set
```
stray comment?

SaravShah replied:
do we need to remove this comment?

jmartin-sul commented:
will leave comments per @ndushay's suggestions (and will get rid of the stray that she caught)

SaravShah commented on spec/models/preserved_copy_spec.rb:
```diff
@@ -309,4 +314,53 @@
     end
   end
+
+  describe '#create_archive_preserved_copies!' do
```
Add 2 tests:
1. ` expect { pc.create_archive_preserved_copies!(pc_version) }.to eq [ ArchivePreservedcopy object]`

2.  `create(:archive_preserved_copy)`
`expect { pc.create_archive_preserved_copies!(pc_version) }.to eq [] `

SaravShah replied:
So we expect that create_archive_preserved_copies returns the archive preserved copies we actually want, AND if we already have the archive preserved copies created, and we call the create_archive_preserved_copies, that method does not return any archive preserved copy objects.

jmartin-sul replied:
right, the method should only create the missing archive copies, and the method should return whatever copies it created.

so, e.g., if no copies were needed, an empty array should be returned.  if a new archive endpoint got added and connected to the relevant preservation policy, the method should create one new archive copy, and should return an instance of that archive copy.

jmartin-sul replied:
#985

jmartin-sul commented:
ok @ndushay and @SaravShah, i believe i've addressed all feedback so far, either by making the requested changes, or by filing and linking issues for follow-on work.  ready for re-review...
